### ISSUE TITLE
[question] Forwarding env_info / buildenv_info to cmake using CMakeDeps

### ISSUE THREAD

**Issue by p-groarke** [conan-io/conan#15079]
### What is your question?

Hello, I am updating a legacy recipe that declared some `env_info` variables. In the past, consuming with `cmake_find_package_multi`, calling a `find_package` on the target would import the variables in cmake.

I am not able to reproduce this anymore. I tried `buildenv_info` and debugging with `VirtualBuildEnv`, I get to the point where the vars are in the `bat` script. But that's as far as I got.

Using conan 1 CMakeDeps (+ CMakeToolchain), how do you forward environment variables from a dependency recipe to the final cmake configure script?

ty!

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by p-groarke** [comment]
OK I've been investigating this and I have a better idea of the CMakeDeps limitations I believe.

In this particular case, we are not calling conan `build`. Conan is used purely to generate config cmake files, and then we use cmake from there. We also don't have access to the cmake toolchain file either, which seems to be where we could add the environment / cmake variables (through `cache_variable`, `preprocessor_definitions`, etc).

Basically, what I'm asking is how to embed environment (cmake) variables in CMakeDeps itself. In the `package-*.cmake` files. Without other means. Without toolchain nor `build` step. Maybe without virtualbuildenv (no clue, that thing is so obscure I can't wrap my head around how to use it hehe).

edit : I think :)

**Comment by jcar87** [comment]
Hi @p-groarke  - thank you for raising this issue.

> We also don't have access to the cmake toolchain file either, 

Could you clarify this?
The idea is that `CMakeDeps` and `CMakeToolchain` generators work in tandem.

Your `conanfile.txt` or `conanfile.py` should use _both_ generators, and then you'd simply pass `-DCMAKE_TOOLCHAIN_FILE=conan_toolchain.cmake` when invoking CMake.  (the location of this file will depend on which layout you are using). 

This is covered in this section of the documentation: https://docs.conan.io/2/tutorial/consuming_packages/use_tools_as_conan_packages.html

I'd encourage you to try that example and visualise the contents of `conan_toolchain.cmake` to see if the relevant variables are there. Otherwise, `conan install` _does_ print the instructions on how to invoke cmake, which you can copy and paste,  in case you can't use presets:

```
======== Finalizing install (deploy, generators) ========
conanfile.txt: Writing generators to /path/to/build
conanfile.txt: Generator 'CMakeToolchain' calling 'generate()'
conanfile.txt: CMakeToolchain generated: conan_toolchain.cmake
conanfile.txt: Preset 'conan-release' added to CMakePresets.json. Invoke it manually using 'cmake --preset conan-release' if using CMake>=3.23
conanfile.txt: If your CMake version is not compatible with CMakePresets (<3.23) call cmake like: 'cmake <path> -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE=/path/to/build/conan_toolchain.cmake -DCMAKE_POLICY_DEFAULT_CMP0091=NEW -DCMAKE_BUILD_TYPE=Release'
conanfile.txt: CMakeToolchain generated: CMakePresets.json
conanfile.txt: Generator 'CMakeDeps' calling 'generate()'
conanfile.txt: Generating aggregated env files
conanfile.txt: Generated aggregated env files: ['conanbuild.sh', 'conanrun.sh']
```


As for anything in `[buildenv]`  - you'd have to call `source conanrun.sh`, which will set up your shell with the relevant environment variables. This shouldn't be much different than how it was with Conan 1.x, other than perhaps the script names - if I remember correctly, the legacy CMake generators in Conan 1.x never actually propagated environment variables.

**Comment by p-groarke** [comment]
@jcar87 Yes of course. We have 2 scenarios. One is a typical conan use case, which does exactly what you mention.

Another is using conan as a dependency provider only. It is called from within the build and is fully decoupled from the cmake invocation. Conan is only used to generate cmake files + import, pretty much.

In that scenario, we cannot pass in the toolchain file and we rely solely on `cmake_find_package_multi` / `CMakeDeps` generators. Other than adding the config files path to cmake prefix, this works great and from what I'm aware of is a supported use case.

My question relates to the second scenario. A decoupled use of conan (without toolchain file). We need to pass along an environment variable as was previously possible with `cmake_find_package_multi`. How does one do that with CMakeDeps only?

Note, if conan doesn't support providing dependencies anymore and requires we tightly couple it to cmake, I doubt we'd continue our enormous task of porting to conan. Or at the very least, we'd probably stick to conan 1 forever (which nobody wants) :(

**Comment by p-groarke** [comment]
Oh and to clarify, `VirtualBuildEnv` won't work either, since (afaik, I'm not an expert), you cannot call a shell script from within the cmake process and inherit its environment variables.

**Comment by memsharded** [comment]
Hi @p-groarke 

> Hello, I am updating a legacy recipe that declared some env_info variables. In the past, consuming with cmake_find_package_multi, calling a find_package on the target would import the variables in cmake.

``cmake_find_package_multi`` does not handle the environment at all. It doesn't even read the environment variables or attributes of dependencies. I think what you saw in the past was due to other means of propagating the environment, but not because of using ``cmake_find_package_multi`` generator.

> We need to pass along an environment variable as was previously possible with cmake_find_package_multi.

I think this was not previously possible. Up to my knowledge this generator (or other cmake generators) have never set up env-vars, that has always happened either in memory (calling Conan commands), or activating environment scripts.

Could you please double check that? It would be good to understand better this issue.

**Comment by p-groarke** [comment]
@memsharded Of course.

In our Qt5 recipe, a simple wrapper around pre-compiled Qt until we port to "the real deal", we use this `package_info`

```python
    def package_info(self):
        # Nothing else, this is enough.
        self.env_info.CMAKE_PREFIX_PATH.append(self.package_folder)
        self.env_info.Qt5_DIR = self.package_folder
```
Then, for some magical reason, cmake is happy once we find the "symbolic" qt package and we can do this :

```cmake
# Import env
find_package(our_custom_qt_recipe CONFIG REQUIRED QUIET)

# Magic!
find_package(Qt5 COMPONENTS Bla Bla REQUIRED QUIET)
```

We are currently shipping this solution and its been working for ~2 years. However, I know nothing and I didn't write this. Scanning our consumer cmake / conanfile, I don't see anything else. So I'm inclined to say find_package is bringing in the env_info vars and everyone is happy.

Note, we recently updated to 1.61, so maybe it's simply a bug?

Edit : This works because our Qt package has all the cmake files required for cmake's built-in find_package.

**Comment by memsharded** [comment]
I am not sure what would be the gap, but I have done the following test:

- Set up a simple recipe with:
```python
    def package_info(self):
        self.env_info.POTATO.append(self.package_folder)
        self.env_info.OTHER_POTATO = self.package_folder
```
- Created a package ``potato/1.0`` for it
- ``require`` that ``potato/1.0`` package, using the ``cmake_find_package_multi``
- The resulting created files do not contain the word POTATO

So it seems impossible that the files generated from ``cmake_find_package_multi`` could be defining anything related to the environment that could help finding the other package.

I suspect that there might be something else, other generator, other script, other mechanism that would make your packages work and define the environment. But not the ``cmake_find_package_multi``.

**Comment by p-groarke** [comment]
The thing is, there really isn't anything else special. I wonder if it was a bug in `cmake_find_package_multi` that allowed this to happen 🤔 I should retest in conan 1.59 which we were using last this worked.

Anyways, this is something direly needed in conan. A way to pass on env / cmake vars without having to resort to toolchain. We need to integrate conan transparently to pre-existing legacy systems. That's the only way for us considering we can't port everything at once and need gradual integration capabilities. [waves hands and yells legacy] XD

The author of our little hack is off on vacation, but I'll ping him once he's back.

**Comment by memsharded** [comment]
I'd also suggest having a look at the problem to solve rather than a specific mechanism. I haven't ever seen any CMake find module or config file to define environment variables, it is likely that it is not the best approach. But there might be other more cmake-native mechanisms to solve your issues. For example the double ``find_package()`` call that you explain above might be solved with CMake built-in variables, so it would be good to know more about that.

Let us know about the above hack when you ping them.

**Comment by AbrilRBS** [comment]
Closing as Conan 1-only, which is no longer supported
